### PR TITLE
Do not fetch activities if they are not enabled

### LINKED ISSUE #788: 2.5.0-rc1 client logs are flooded by 404 error
I´m using Nextcloud 14 + Nextcloud desktop client 2.5 RC1 for Windows. I upgrade from a previous version of the client and if i open the `--logwindow` the logs are flooded with the following lines:
```
[OCC::ActivityListModel::startFetchJob 	Start fetching activities for  "sombody@cloud.xxxxxxxx.com"
[OCC::AccessManager::createRequest 	2 "" "https://cloud.xxxxxxxx.com/ocs/v2.php/cloud/activity?page=0&pagesize=100&format=json" has X-Request-ID "XXXXXXXX-XXXX-XXXX-XXXX-XXXXXXXXXXXX"
[OCC::AbstractNetworkJob::start 	OCC::JsonApiJob created for "https://cloud.xxxxxxxx.com" + "ocs/v2.php/cloud/activity" "OCC::ActivityListModel"
[OCC::AbstractNetworkJob::slotFinished 	QNetworkReply::NetworkError(ContentNotFoundError) "Error transferring https://cloud.xxxxxxxx.com/ocs/v2.php/cloud/activity?page=0&pagesize=100&format=json - server replied: " QVariant(int, 404)
[OCC::JsonApiJob::finished 	JsonApiJob of QUrl("https://cloud.xxxxxxxx.com/ocs/v2.php/cloud/activity?page=0&pagesize=100&format=json") FINISHED WITH STATUS "ContentNotFoundError Error transferring https://cloud.xxxxxxxx.com/ocs/v2.php/cloud/activity?page=0&pagesize=100&format=json - server replied: "
[OCC::JsonApiJob::finished 	Network error:  "ocs/v2.php/cloud/activity" "Error transferring https://cloud.xxxxxxxx.com/ocs/v2.php/cloud/activity?page=0&pagesize=100&format=json - server replied: " QVariant(int, 404)
```
Every one or two seconds new lines pop up with a different request id.

### LINKED ISSUE #834: Server flooded with requests from 2.5.0 client (activity)
Version 2.5.0 of the Nextcloud client bombs the server with requests to `/ocs/v2.php/cloud/activity?page=0&pagesize=100&format=json`.

I am seeing up to **16 requests per second per user** with the new client, this is way too much. The new client can bring down a home server easily with just a few users. Even in larger environments this can be problematic, as there are potentially even more users.

The strange part is that not all 2.5.0 clients cause this amount of traffic. Some clients (including the one I am using) only hit the activity endpoint once a minute.

First of all, this is the relevant client identifier from the server log:

```
Mozilla/5.0 (Windows) mirall/2.5.0v2.5.0 (build 20181112) (Nextcloud)
```

After locking out this one client (return 403 for this source IP) the client hit the server with **150 requests/second**.

I can tell for sure that this is not an attack, it is definitely a faulty client. Maybe someone could check the code path for polling the activity log how this can happen.

### Expected behaviour

The client should only poll once a minute (like in older releases).

### Actual behaviour

_Some_ 2.5.0 clients poll without sleeping in between, up to 150 times a second in my case.

### Steps to reproduce

I am not sure, as I do not have that particular client under my control.

### Client configuration

Client version: 2.5.0

Operating system: Windows

### Server configuration

Operating system: CentOS 7.5

Web server: Apache 2.4.6-80

Database: Postgres

PHP version:7.2.12

Nextcloud version: 14.0.3

Storage backend (external storage): -

### Logs

1. Client logfile: Output of `nextcloud --logwindow` or `nextcloud --logfile log.txt`
(On Windows using `cmd.exe`, you might need to first `cd` into the Nextcloud directory)
(See also https://docs.nextcloud.com/desktop/2.3/troubleshooting.html#log-files)

2. Web server access log:
```
192.168.12.9 - user12 [15/Nov/2018:14:56:07 +0100] "GET /ocs/v2.php/cloud/activity?page=0&pagesize=100&format=json HTTP/1.1" 200 74 "-" "Mozilla/5.0 (Windows) mirall/2.5.0v2.5.0 (build 20181112) (Nextcloud)" 0
192.168.12.9 - user12 [15/Nov/2018:14:56:08 +0100] "GET /ocs/v2.php/cloud/activity?page=0&pagesize=100&format=json HTTP/1.1" 200 74 "-" "Mozilla/5.0 (Windows) mirall/2.5.0v2.5.0 (build 20181112) (Nextcloud)" 0
192.168.12.9 - user12 [15/Nov/2018:14:56:08 +0100] "GET /ocs/v2.php/cloud/activity?page=0&pagesize=100&format=json HTTP/1.1" 200 74 "-" "Mozilla/5.0 (Windows) mirall/2.5.0v2.5.0 (build 20181112) (Nextcloud)" 0
192.168.12.9 - user12 [15/Nov/2018:14:56:08 +0100] "GET /ocs/v2.php/cloud/activity?page=0&pagesize=100&format=json HTTP/1.1" 200 74 "-" "Mozilla/5.0 (Windows) mirall/2.5.0v2.5.0 (build 20181112) (Nextcloud)" 0
192.168.12.9 - user12 [15/Nov/2018:14:56:08 +0100] "GET /ocs/v2.php/cloud/activity?page=0&pagesize=100&format=json HTTP/1.1" 200 74 "-" "Mozilla/5.0 (Windows) mirall/2.5.0v2.5.0 (build 20181112) (Nextcloud)" 0
192.168.12.9 - user12 [15/Nov/2018:14:56:08 +0100] "GET /ocs/v2.php/cloud/activity?page=0&pagesize=100&format=json HTTP/1.1" 200 74 "-" "Mozilla/5.0 (Windows) mirall/2.5.0v2.5.0 (build 20181112) (Nextcloud)" 0
192.168.12.9 - user12 [15/Nov/2018:14:56:08 +0100] "GET /ocs/v2.php/cloud/activity?page=0&pagesize=100&format=json HTTP/1.1" 200 74 "-" "Mozilla/5.0 (Windows) mirall/2.5.0v2.5.0 (build 20181112) (Nextcloud)" 0
192.168.12.9 - user12 [15/Nov/2018:14:56:08 +0100] "GET /ocs/v2.php/cloud/activity?page=0&pagesize=100&format=json HTTP/1.1" 200 74 "-" "Mozilla/5.0 (Windows) mirall/2.5.0v2.5.0 (build 20181112) (Nextcloud)" 0
192.168.12.9 - user12 [15/Nov/2018:14:56:08 +0100] "GET /ocs/v2.php/cloud/activity?page=0&pagesize=100&format=json HTTP/1.1" 200 74 "-" "Mozilla/5.0 (Windows) mirall/2.5.0v2.5.0 (build 20181112) (Nextcloud)" 0
192.168.12.9 - user12 [15/Nov/2018:14:56:08 +0100] "GET /ocs/v2.php/cloud/activity?page=0&pagesize=100&format=json HTTP/1.1" 200 74 "-" "Mozilla/5.0 (Windows) mirall/2.5.0v2.5.0 (build 20181112) (Nextcloud)" 0
192.168.12.9 - user12 [15/Nov/2018:14:56:08 +0100] "GET /ocs/v2.php/cloud/activity?page=0&pagesize=100&format=json HTTP/1.1" 200 74 "-" "Mozilla/5.0 (Windows) mirall/2.5.0v2.5.0 (build 20181112) (Nextcloud)" 0
192.168.12.9 - user12 [15/Nov/2018:14:56:08 +0100] "GET /ocs/v2.php/cloud/activity?page=0&pagesize=100&format=json HTTP/1.1" 200 74 "-" "Mozilla/5.0 (Windows) mirall/2.5.0v2.5.0 (build 20181112) (Nextcloud)" 0
192.168.12.9 - user12 [15/Nov/2018:14:56:08 +0100] "GET /ocs/v2.php/cloud/activity?page=0&pagesize=100&format=json HTTP/1.1" 200 74 "-" "Mozilla/5.0 (Windows) mirall/2.5.0v2.5.0 (build 20181112) (Nextcloud)" 0
192.168.12.9 - user12 [15/Nov/2018:14:56:08 +0100] "GET /ocs/v2.php/cloud/activity?page=0&pagesize=100&format=json HTTP/1.1" 200 74 "-" "Mozilla/5.0 (Windows) mirall/2.5.0v2.5.0 (build 20181112) (Nextcloud)" 0
192.168.12.9 - user12 [15/Nov/2018:14:56:08 +0100] "GET /ocs/v2.php/cloud/activity?page=0&pagesize=100&format=json HTTP/1.1" 200 74 "-" "Mozilla/5.0 (Windows) mirall/2.5.0v2.5.0 (build 20181112) (Nextcloud)" 0
192.168.12.9 - user12 [15/Nov/2018:14:56:08 +0100] "GET /ocs/v2.php/cloud/activity?page=0&pagesize=100&format=json HTTP/1.1" 200 74 "-" "Mozilla/5.0 (Windows) mirall/2.5.0v2.5.0 (build 20181112) (Nextcloud)" 0
192.168.12.9 - user12 [15/Nov/2018:14:56:08 +0100] "GET /ocs/v2.php/cloud/activity?page=0&pagesize=100&format=json HTTP/1.1" 200 74 "-" "Mozilla/5.0 (Windows) mirall/2.5.0v2.5.0 (build 20181112) (Nextcloud)" 0
192.168.12.9 - user12 [15/Nov/2018:14:56:09 +0100] "GET /ocs/v2.php/cloud/activity?page=0&pagesize=100&format=json HTTP/1.1" 200 74 "-" "Mozilla/5.0 (Windows) mirall/2.5.0v2.5.0 (build 20181112) (Nextcloud)" 0
192.168.12.9 - user12 [15/Nov/2018:14:56:09 +0100] "GET /ocs/v2.php/cloud/activity?page=0&pagesize=100&format=json HTTP/1.1" 200 74 "-" "Mozilla/5.0 (Windows) mirall/2.5.0v2.5.0 (build 20181112) (Nextcloud)" 0

```

### PR DESCRIPTION
**Pull request by rullzer**
Fixes #788
Fixes #834

1. Disable activities app on the server
2. Open client
3. Open activities tab

Before:
- observe a :hankey:-ton of requests to your server

After:
- No :hankey:-ton of requests to your server

### CODE REVIEW

**Comment by jpnurmi** [comment]
I can still reproduce the problem with a test account I created on webo.hosting, with activities enabled. :( On the first run, when I add the account, the flood starts when I scroll to the bottom of the activity list. On the second run, the flood starts immediately on startup and the activity list is empty.

Server v14.0.3.0:
```
[OCC::CheckServerJob::finished 	status.php returns:  QJsonDocument({"edition":"","installed":true,"maintenance":false,"needsDbUpgrade":false,"productname":"Nextcloud by Webo.hosting","version":"14.0.3.0","versionstring":"14.0.3"})   QNetworkReply::NetworkError(NoError)  Reply:  QNetworkReplyHttpImpl(0x555555fdac50)
```

Has activities capability:
```
[OCC::JsonApiJob::finished 	JsonApiJob of QUrl("https://nextcloud-fi.webo.hosting/ocs/v1.php/cloud/capabilities?format=json") FINISHED WITH STATUS "OK"
[OCC::ConnectionValidator::slotCapabilitiesRecieved 	Server capabilities QJsonObject({"activity":{"apiv2":["filters","filters-api","previews","rich-strings"]},"bruteforce":{"delay":0},"core":{"pollinterval":60,"webdav-root":"remote.php/webdav"},"dav":{"chunking":"1.0"},"external":{"v1":["sites","device","groups","redirect"]},"files":{"bigfilechunking":true,"blacklisted_files":[".htaccess"],"undelete":true,"versioning":true},"files_sharing":{"api_enabled":true,"default_permissions":31,"federation":{"expire_date":{"enabled":true},"incoming":true,"outgoing":true},"group":{"enabled":false,"expire_date":{"enabled":true}},"group_sharing":false,"public":{"enabled":true,"expire_date":{"enabled":false},"password":{"enforced":false},"send_mail":false,"upload":true,"upload_files_drop":true},"resharing":true,"sharebymail":{"enabled":true,"expire_date":{"enabled":true},"password":{"enabled":true},"upload_files_drop":{"enabled":true}},"user":{"expire_date":{"enabled":true},"send_mail":false}},"fulltextsearch":{"providers":[],"remote":true},"notifications":{"admin-notifications":["ocs","cli"],"ocs-endpoints":["list","get","delete","delete-all","icons","rich-strings"],"push":["devices","object-data"]},"ocm":{"apiVersion":"1.0-proposal1","enabled":true,"endPoint":"https://nextcloud-fi.webo.hosting/index.php/ocm","shareTypes":[{"name":"file","protocols":{"webdav":"/public.php/webdav/"}}]},"password_policy":{"enforceNonCommonPassword":true,"enforceNumericCharacters":true,"enforceSpecialCharacters":true,"enforceUpperLowerCase":true,"minLength":8},"richdocuments":{"mimetypes":["application/vnd.oasis.opendocument.text","application/vnd.oasis.opendocument.spreadsheet","application/vnd.oasis.opendocument.presentation","application/vnd.openxmlformats-officedocument.wordprocessingml.document","application/vnd.openxmlformats-officedocument.spreadsheetml.sheet","application/vnd.openxmlformats-officedocument.presentationml.presentation"]},"spreed":{"features":["audio","video","chat-v2","guest-signaling","empty-group-room","guest-display-names","multi-room-users","favorites","last-room-activity","no-ping","system-messages","mention-flag","in-call-flags"]},"theming":{"background":"https://nextcloud-fi.webo.hosting/core/img/background.png?v=5","background-default":true,"background-plain":false,"color":"#0082c9","color-element":"#0082c9","color-text":"#ffffff","logo":"https://nextcloud-fi.webo.hosting/index.php/apps/theming/image/logo?useSvg=1&v=5","name":"Nextcloud by Webo.hosting","slogan":"a safe home for all your data","url":"https://webo.hosting/en/webo-cloud-en/"}})
```

**Comment by rullzer** [comment]
Mmmm so another case it seems.
Let me check that out thnx for testing

**Comment by jpnurmi** [comment]
The loop looks like this:
```
ActivityListModel::slotRefreshActivity()
-> ActivityListModel::startFetchJob()
-> ActivityListModel::slotActivitiesReceived()
-> ActivityListModel::combineActivityLists()
-> QAbstractItemModel::endInsertRows()
-> QAbstractItemView::rowsInserted() // fetchMoreTimer.start()
-> QAbstractItemView::timerEvent()
-> QAbstractItemViewPrivate::fetchMore()
-> ActivityListModel::fetchMore()
-> ActivityListModel::startFetchJob()
-> ActivityListModel::slotActivitiesReceived()
...
```

**Comment by jpnurmi** [comment]
The loop can be easily broken by _not_ calling `begin/EndResetModel()` + `begin/endInsertRows()` _if_ `_finalList == resultList`.

File: src/gui/activitylistmodel.cpp [hunk 1]
```diff
@@ -291,18 +291,28 @@ void ActivityListModel::combineActivityLists()
...
 
+bool ActivityListModel::canFetchActivities() const {
+    return _accountState->isConnected() && _accountState->account()->capabilities().hasActivities();
+}
+
 void ActivityListModel::fetchMore(const QModelIndex &)
 {
-    if (_accountState->isConnected()) {
-        _activityLists = ActivityList();
+    _activityLists = ActivityList();
```
Comment (jpnurmi) [inline src/gui/activitylistmodel.cpp]: i'm quite confused why fetchMore() clears previously fetched activities

Comment (rullzer) [reply src/gui/activitylistmodel.cpp]: That is actually a good question

**Pull request merged**